File: 백준/Gold/7576. 토마토/토마토.py
from collections import deque


def bfs():
    queue = deque()

    for i in range(m):
        for j in range(n):
            if graph[i][j] == 1:
                queue.append((i, j))

    while queue:
        x, y = queue.popleft()

        for i in range(4):
            nx = dx[i] + x
            ny = dy[i] + y

            if 0 <= nx < m and 0 <= ny < n:
                if graph[nx][ny] == 0:
                    queue.append((nx, ny))
                    graph[nx][ny] = graph[x][y] + 1


n, m = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(m)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

visited = [[False] * n for _ in range(m)]


# 저장될 때부터 다 익어있는지 확인
flag = True  # 다 익음이 디폴트
for i in range(m):
    for j in range(n):
        if graph[i][j] == 0:
            flag = False  # 다 익어있지않음

if flag:
    print(0)
    exit()

# 시작 좌표를 주지 않는 이유
# 주면 그 좌포에서 상하좌우로 퍼지기 시작하는데
# 다른 익은 토마토가 있는 지점에서도 동시에 퍼지기 시작해야됨
bfs()

result = 0
for i in range(m):
    for j in range(n):
        if graph[i][j] == 0:
            print(-1)
            exit()
        result = max(result, graph[i][j])

print(result - 1)

File: 백준/Gold/7576. 토마토/토마토.java
import java.io.*;
import java.util.*;

public class Main {
    
    static int n, m;
    static int[][] graph;
    static final int[] dx = {-1 , 1, 0, 0};
    static final int[] dy = {0, 0, -1, 1};

    private static int bfs() {
        ArrayDeque<int[]> queue = new ArrayDeque<>();
        
        // 익은 토마토를 큐에 모두 삽입
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (graph[i][j] == 1) {
                    queue.offer(new int[]{i, j});
                }
            }
        }

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int x = cur[0];
            int y = cur[1];

            for (int i = 0; i < 4; i++) {
                int nx = dx[i] + x;
                int ny = dy[i] + y;

                if ((0 <= nx && nx < n) && (0 <= ny && ny < m)) {
                    if (graph[nx][ny] == 0) {
                        graph[nx][ny] = graph[x][y] + 1;
                        queue.offer(new int[]{nx, ny});
                    }
                }
            }
        }

        int maxVal = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (graph[i][j] == 0) {
                    return -1;
                }
                maxVal = Math.max(maxVal, graph[i][j]);
            }
        }

        return maxVal - 1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());

        graph = new int[n][m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int result = bfs();
        bw.write(result + "\n");
        bw.flush();
    }
}   
